lib/srv/reexec.go

```diff
@@ -612,9 +612,96 @@ func (o *osWrapper) startNewParker(ctx context.Context, credential *syscall.Cred
 	return nil
 }
 
-// RunForward reads in the command to run from the parent process (over a
+type forwardHandlerParams struct {
+	addr           string
+	fileDescriptor *os.File
+}
+
+type forwardHandler func(ctx context.Context, params forwardHandlerParams) error
+
+func handleLocalPortForward(ctx context.Context, params forwardHandlerParams) error {
+	if params.fileDescriptor == nil {
+		return trace.BadParameter("parent process did not send a socket")
+	}
+	conn, err := uds.FromFile(params.fileDescriptor)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if err := params.fileDescriptor.Close(); err != nil {
+		return trace.Wrap(err)
+	}
+
+	go func(addr string, conn net.Conn) {
+		defer conn.Close()
+		var d net.Dialer
+		remote, err := d.DialContext(ctx, "tcp", addr)
+		if err != nil {
+			log.WithError(err).Errorf("error dialing %v", addr)
+		}
+		defer remote.Close()
+		if err := utils.ProxyConn(ctx, conn, remote); err != nil {
+			log.WithError(err).Errorf("error proxying connection to %v", addr)
+		}
+	}(params.addr, conn)
+
+	return nil
+}
+
+func createRemotePortForwardingListener(ctx context.Context, params forwardHandlerParams) (*os.File, error) {
+	var lc net.ListenConfig
+	listener, err := lc.Listen(ctx, "tcp", params.addr)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	defer listener.Close()
+
+	var tcpListener *net.TCPListener
+	var ok bool
+	if tcpListener, ok = listener.(*net.TCPListener); !ok {
+		return nil, trace.Errorf("expected listener to be of type *net.TCPListener, but was %T", listener)
+	}
+
+	listenerFD, err := tcpListener.File()
+	return listenerFD, trace.Wrap(err)
+}
+
+func handleRemotePortForward(ctx context.Context, params forwardHandlerParams) error {
```

Comment:
This will block the parent control connection until the listener is established and sent back successfully; we should run handlers in a goroutine in `runForward` (so `handleLocalPortForward` won't need to spawn its own goroutine).